funsor/sum_product.py

```diff
@@ -251,6 +252,149 @@ def partial_sum_product(
     return results
 
 
+def dynamic_partial_sum_product(
+    sum_op, prod_op, factors, eliminate=frozenset(), plate_to_step=dict()
+):
```

Comment:
After some thought, for this PR it is probably better to keep this interface since `plate_to_step` here is consistent with `pyro.vectorized_markov` and existing unit tests, thus, requires least amount of change. And defer possible interface changes to the follow up PR (#539). WDYT?